### PR TITLE
Refactoring adaptation

### PR DESCRIPTION
**Pull request by xukai92**

### CODE REVIEW

File: src/samplers/nuts.jl [hunk 1]
```diff
@@ -192,7 +192,7 @@ function _nuts_step(θ::T, ϵ::AbstractFloat, lj_func::Function, grad_func::Func
                     j_max::Int=5) where {T<:Union{AbstractVector,SubArray}}
 
   d = length(θ)
-  r0 = randn(d)
+  r0 = _sample_momentum(d, stds)
```
Comment (xukai92) [inline src/samplers/nuts.jl]: @yebai Actually there was no bug in adaptation. This is the line which leads to the bug in our NUTS.

**Comment by xukai92** [comment]
@yebai Can you take a look over the new adaptation codes? I think it's already clear enough (even clear than other references we were looking at) and easy-to-extend (e.g. adding a covariance pre-cond adapter).

**Comment by mohamed82008** [comment]
`@TuringBot benchmark`

File: src/Turing.jl [hunk 2]
```diff
@@ -88,9 +88,12 @@ const CACHERANGES = 0b01
 #######################
 # Sampler abstraction #
 #######################
-
+abstract type AbstractState end
```
Comment (yebai) [inline src/Turing.jl]: AbstractState ==> AbstractAdaptState?

**Comment by yebai** [comment]
“Naming consistency”
==> Adapt == > Adaptor?

“Stan”  

==> move all `cmdstan` related code to `adapt/stan.jl` 


“AbstractState”

==> perhaps remove “AbstractState” since it’s not really helpful. 


“DPCState + FPCState + NullPC”

==> AbstractPreConditioner —> UnitPreConditioner + DensePreConditioner + DiagPreConditioner

**Comment by xukai92** [comment]
- [ ] improve naming
- [ ] put Stan related codes in a single file
- [ ] remove unnecessary state types

**Comment by xukai92** [comment]
@yebai I addressed the comments you made.

**Pull request merged**